File: api/utils/validation.js
const { spawnSync } = require('child_process');
const { SAFETY_LIMITS } = require('../config/constants');

function validateVideoFile(filePath) {
  try {
    const result = spawnSync('ffprobe', [
      '-v', 'error',
      '-select_streams', 'v',
      '-show_entries', 'stream=codec_type',
      '-of', 'csv=p=0',
      filePath
    ], { encoding: 'utf8', timeout: 10000 });
    if (result.status !== 0 || result.error) return false;
    return (result.stdout || '').trim().includes('video');
  } catch {
    return false;
  }
}

function validateTimeParam(value) {
  if (!value) return null;
  const str = String(value).trim();
  if (!str) return null;

  if (/^\d+(\.\d+)?$/.test(str)) {
    const num = parseFloat(str);
    return isFinite(num) && num >= 0 ? str : null;
  }

  const hhmmss = /^(?:(\d{1,2}):)?(\d{1,2}):(\d{1,2}(?:\.\d+)?)$/.exec(str);
  if (hhmmss) {
    const h = hhmmss[1] ? parseInt(hhmmss[1], 10) : 0;
    const m = parseInt(hhmmss[2], 10);
    const s = parseFloat(hhmmss[3]);
    if (m < 60 && s < 60 && isFinite(h) && isFinite(m) && isFinite(s)) {
      return str;
    }
  }
  return null;
}

function validateUrl(url) {
  if (!url || typeof url !== 'string') {
    return { valid: false, error: 'URL is required' };
  }

  if (url.length > SAFETY_LIMITS.maxUrlLength) {
    return { valid: false, error: 'URL is too long' };
  }

  try {
    const parsed = new URL(url);

    if (!['http:', 'https:'].includes(parsed.protocol)) {
      return { valid: false, error: 'Only HTTP/HTTPS URLs are allowed' };
    }

    const hostname = parsed.hostname.toLowerCase();
    const privatePatterns = [
      /^localhost$/i,
      /^127\./,
      /^10\./,
      /^172\.(1[6-9]|2[0-9]|3[01])\./,
      /^192\.168\./,
      /^0\./,
      /^169\.254\./,
      /^\[::1\]$/,
      /^\[fe80:/i,
      /^\[fc00:/i,
      /^\[fd00:/i
    ];

    for (const pattern of privatePatterns) {
      if (pattern.test(hostname)) {
        return { valid: false, error: 'Private/local URLs are not allowed' };
      }
    }

    return { valid: true };
  } catch (e) {
    return { valid: false, error: 'Invalid URL format' };
  }
}

module.exports = {
  validateVideoFile,
  validateTimeParam,
  validateUrl
};

File: internal/util/validation.go
package util

import (
	"net"
	"net/url"
	"os/exec"
	"regexp"
	"strings"

	"github.com/coah80/yoink/internal/config"
)

var timeRe = regexp.MustCompile(`^(?:(\d{1,2}):)?(\d{1,2}):(\d{1,2}(?:\.\d+)?)$`)
var numRe = regexp.MustCompile(`^\d+(\.\d+)?$`)

type URLValidation struct {
	Valid bool
	Error string
}

func ValidateURL(rawURL string) URLValidation {
	if rawURL == "" {
		return URLValidation{false, "URL is required"}
	}
	if len(rawURL) > config.MaxURLLength {
		return URLValidation{false, "URL is too long"}
	}

	parsed, err := url.Parse(rawURL)
	if err != nil {
		return URLValidation{false, "Invalid URL format"}
	}

	if parsed.Scheme != "http" && parsed.Scheme != "https" {
		return URLValidation{false, "Only HTTP/HTTPS URLs are allowed"}
	}

	hostname := strings.ToLower(parsed.Hostname())
	if isPrivateHost(hostname) {
		return URLValidation{false, "Private/local URLs are not allowed"}
	}

	return URLValidation{true, ""}
}

var privateNets []*net.IPNet

func init() {
	cidrs := []string{
		"127.0.0.0/8",
		"10.0.0.0/8",
		"172.16.0.0/12",
		"192.168.0.0/16",
		"0.0.0.0/8",
		"169.254.0.0/16",
		"::1/128",
		"fe80::/10",
		"fc00::/7",
	}
	for _, cidr := range cidrs {
		_, network, _ := net.ParseCIDR(cidr)
		privateNets = append(privateNets, network)
	}
}

func isPrivateIP(ip net.IP) bool {
	for _, network := range privateNets {
		if network.Contains(ip) {
			return true
		}
	}
	return false
}

func isPrivateHost(hostname string) bool {
	if hostname == "" || hostname == "localhost" {
		return true
	}

	ip := net.ParseIP(hostname)
	if ip == nil {
		ip = net.ParseIP(strings.Trim(hostname, "[]"))
	}

	if ip != nil {
		return isPrivateIP(ip)
	}

	ips, err := net.LookupIP(hostname)
	if err != nil {
		return true
	}
	for _, ip := range ips {
		if isPrivateIP(ip) {
			return true
		}
	}
	return false
}

func ValidateVideoFile(filePath string) bool {
	cmd := exec.Command("ffprobe",
		"-v", "error",
		"-select_streams", "v",
		"-show_entries", "stream=codec_type",
		"-of", "csv=p=0",
		filePath,
	)
	out, err := cmd.Output()
	if err != nil {
		return false
	}
	return strings.Contains(string(out), "video")
}

func ValidateTimeParam(value string) string {
	value = strings.TrimSpace(value)
	if value == "" {
		return ""
	}

	if numRe.MatchString(value) {
		return value
	}

	if timeRe.MatchString(value) {
		return value
	}
	return ""
}
